kvserver: group fields mutated under raftMu+mu

pav-kv opened the pull request:
This PR groups fields mutated under both `raftMu+mu` into a section of `Replica` fields explicitly marked/commented as such. All the fields in this section can be read with only one of the two mutexes held.

This PR also eliminates a few cases when `Replica.mu` is locked unnecessarily.

Epic: none
Release note: none

pav-kv committed:
kvserver: group fields mutated under both raftMu/mu

Epic: none
Release note: none

pav-kv committed:
kvserver: remove unnecessary locking

Epic: none
Release note: none

pav-kv committed:
kvserver: include replica state to orRaftMu

Epic: none
Release note: none

pav-kv committed:
kvserver: remove unnecessary locking

Epic: none
Release note: none

pav-kv committed:
kvserver: rm unnecessary RLock in GetSnapshot

Epic: none
Release note: none

pav-kv committed:
kvserver: move raftMu+mu vars to shared section

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131493)
<!-- Reviewable:end -->

pav-kv commented:
@tbg I'm on the fence w.r.t. naming/placing this section. Initially I placed it in `mu.orRaftMu` sub-section, but then it looked unnecessarily verbose so I moved it to a dedicated `shMu` section.

pav-kv commented on pkg/kv/kvserver/replica.go:
```diff
@@ -436,6 +436,46 @@ type Replica struct {
+	// held. They can be accessed when either of the two mutexes is held.
+	//
+	// TODO(pav-kv): audit all other fields and include here.
+	shMu struct {
```
```
"Do not communicate by sharing memory; instead, share memory by communicating."
```

tbg replied:
Much better! Thank you for slogging through another mass rename. I like where you got this.

tbg replied:
Not going to review this commit line-by-line on the assumption that it's 100% mechanical - let me know if there are parts that could use a human check.

pav-kv replied:
The big commits are 100% mechanical. The "interesting" ones (where I removed unnecessary locking) are separate and small, so hopefully you reviewed them.

pav-kv commented:
There is apparently some deadlock in CI, might have accidentally caused by a few `Lock`s added in tests to maintain the fact that these fields must be modified while both mutexes are held.

tbg commented on pkg/kv/kvserver/replica.go:
```diff
@@ -1289,7 +1288,9 @@ func (r *Replica) entireSpanExcludedFromBackupRLocked(
 
 // Version returns the replica version.
 func (r *Replica) Version() roachpb.Version {
-	if r.mu.state.Version == nil {
```
oh, this was a bug, right?

tbg commented on pkg/kv/kvserver/replica_closedts_internal_test.go:
```diff
@@ -565,10 +565,12 @@ func TestReplicaClosedTimestamp(t *testing.T) {
 			cfg.TestingKnobs.DontCloseTimestamps = true
 			cfg.ClosedTimestampReceiver = &r
 			tc.StartWithStoreConfig(ctx, t, stopper, cfg)
+			tc.repl.raftMu.Lock()
```
Hmm, in light of the need to hold both mutexes, `mu.orRaftMu` is a deceptive name. (Disregard if this changes by the end of the commit list).

tbg commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -405,6 +405,8 @@ func (r *Replica) makeReproposal(origP *ProposalData) (reproposal *ProposalData,
 }
 
 func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *replicatedCmd) error {
```
Rename with `raftMuLocked` suffix? Let's stick to the naming conventions unless there's a strong reason not to.

tbg reviewed (approved):
LGTM and thanks. Let's get this in before it rots.

pav-kv committed:
kvserver: fix locking in SetMVCCStatsForTesting

This is a test-only method used in one test, so the bug wasn't critical.

Epic: none
Release note: none

pav-kv commented:
TFTR!

bors r=tbg
